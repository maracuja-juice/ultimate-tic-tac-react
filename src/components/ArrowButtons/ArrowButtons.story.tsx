import * as React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { action } from '@storybook/addon-actions';
import { IconProvider } from '../Icons';
import { ArrowButtons, ArrowButtonsProps } from './ArrowButtons';

export default {
  title: 'ArrowButtons',
  component: ArrowButtons,
} as Meta;

export const Default: Story<ArrowButtonsProps> = (args) => (
  <IconProvider>
    <div className="historyButtons">
      <ArrowButtons
        {...args}
        onInteraction={(forward) => action('interaction: ' + forward)}
      />
    </div>
  </IconProvider>
);
Default.args = {
  value: 1,
  minValue: 1,
  maxValue: 5,
};
