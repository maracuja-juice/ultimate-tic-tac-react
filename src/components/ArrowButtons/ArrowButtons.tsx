import * as React from 'react';
import { Button, ButtonProps } from '@rmwc/button';
import { ArrowLeftIcon, ArrowRightIcon } from '../Icons';
import styles from './ArrowButtons.module.css';
import { useEffect } from 'react';

export interface ButtonConfig {
  buttonProps?: ButtonProps;
  hide?: boolean;
}

export interface ArrowButtonsProps {
  value: number;
  maxValue: number;
  minValue: number;
  onInteraction: (forward: boolean) => void;
  handleKeyboard?: boolean;
  leftButtonConfig?: ButtonConfig;
  rightButtonConfig?: ButtonConfig;
  children?: React.ReactNode;
}

const handleInteraction = (
  forward: boolean,
  value: number,
  minValue: number,
  maxValue: number,
  handle: (f: boolean) => void,
) => {
  if ((forward && value === maxValue) || (!forward && value === minValue))
    return;

  handle(forward);
};

export function ArrowButtons(props: ArrowButtonsProps) {
  const {
    handleKeyboard,
    value,
    maxValue,
    minValue,
    onInteraction,
    leftButtonConfig,
    rightButtonConfig,
    children,
  } = props;
  useEffect(() => {
    const handleEvent = (event: KeyboardEvent) => {
      if (event.key === 'ArrowLeft' || event.key === 'Left')
        handleInteraction(false, value, minValue, maxValue, onInteraction);
      else if (event.key === 'ArrowRight' || event.key === 'Right')
        handleInteraction(true, value, minValue, maxValue, onInteraction);
    };

    if (handleKeyboard) window.addEventListener('keydown', handleEvent);
    return () => {
      if (handleKeyboard) window.removeEventListener('keydown', handleEvent);
    };
  }, [handleKeyboard, value, minValue, maxValue, onInteraction]);

  return (
    <>
      {!leftButtonConfig?.hide && (
        <Button
          disabled={value === minValue}
          dense={true}
          raised={true}
          icon={{ icon: <ArrowLeftIcon />, 'aria-hidden': true }}
          onClick={() => {
            handleInteraction(false, value, minValue, maxValue, onInteraction);
          }}
          className={styles.buttonMargin}
          {...leftButtonConfig?.buttonProps}
        >
          Previous
        </Button>
      )}
      {children}
      {!rightButtonConfig?.hide && (
        <Button
          disabled={value === maxValue}
          dense={true}
          raised={true}
          trailingIcon={{ icon: <ArrowRightIcon />, 'aria-hidden': true }}
          onClick={() => {
            handleInteraction(true, value, minValue, maxValue, onInteraction);
          }}
          className={styles.buttonMargin}
          {...rightButtonConfig?.buttonProps}
        >
          Next
        </Button>
      )}
    </>
  );
}
