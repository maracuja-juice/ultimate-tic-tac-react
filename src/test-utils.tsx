// See: https://testing-library.com/docs/react-testing-library/setup#custom-render
import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { IconProvider } from './components/Icons';
import '@testing-library/jest-dom'; // https://github.com/testing-library/jest-dom/issues/282#issuecomment-660034778
import { MemoryRouter } from 'react-router-dom';

const AllTheProviders: FC = ({ children }) => {
  return (
    <IconProvider>
      <MemoryRouter>{children}</MemoryRouter>
    </IconProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
