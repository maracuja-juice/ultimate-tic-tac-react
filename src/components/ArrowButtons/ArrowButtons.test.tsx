import * as React from 'react';
import { render, screen } from '../../test-utils';
import userEvent, { specialChars } from '@testing-library/user-event';
import { ArrowButtons } from './ArrowButtons';

test('renders next & previous buttons', () => {
  render(
    <ArrowButtons
      value={2}
      minValue={1}
      maxValue={5}
      onInteraction={() => {}}
    />,
  );
  const next = screen.getByRole('button', { name: /next/i });
  const previous = screen.getByRole('button', { name: /previous/i });
  expect(next).toBeInTheDocument();
  expect(next).not.toBeDisabled();
  expect(previous).toBeInTheDocument();
  expect(previous).not.toBeDisabled();
});

test('disables next button when value is maxValue', () => {
  const onInteraction = jest.fn();
  render(
    <ArrowButtons
      value={1}
      minValue={0}
      maxValue={1}
      onInteraction={onInteraction}
    />,
  );

  const next = screen.getByRole('button', { name: /next/i });
  userEvent.click(next);
  expect(next).toBeDisabled();
  expect(onInteraction).not.toHaveBeenCalled();
});

test('disables previous button when value is minValue', () => {
  const onInteraction = jest.fn();
  render(
    <ArrowButtons
      value={0}
      minValue={0}
      maxValue={1}
      onInteraction={onInteraction}
    />,
  );

  const previous = screen.getByRole('button', { name: /previous/i });
  userEvent.click(previous);
  expect(previous).toBeDisabled();
  expect(onInteraction).not.toHaveBeenCalled();
});

test('allows clicking next button', () => {
  const onInteraction = jest.fn();
  render(
    <ArrowButtons
      value={2}
      minValue={1}
      maxValue={5}
      onInteraction={onInteraction}
    />,
  );

  const next = screen.getByRole('button', { name: /next/i });
  userEvent.click(next);

  expect(onInteraction).toHaveBeenCalledWith(true);
});

test('allows clicking previous button', () => {
  const onInteraction = jest.fn();
  render(
    <ArrowButtons
      value={2}
      minValue={1}
      maxValue={5}
      onInteraction={onInteraction}
    />,
  );

  const previous = screen.getByRole('button', { name: /previous/i });
  userEvent.click(previous);

  expect(onInteraction).toHaveBeenCalledWith(false);
});

describe('handleKeyboard', () => {
  test('allows using arrow keys', () => {
    const onInteraction = jest.fn();
    render(
      <ArrowButtons
        value={2}
        minValue={1}
        maxValue={5}
        onInteraction={onInteraction}
        handleKeyboard={true}
      />,
    );

    userEvent.keyboard(specialChars.arrowLeft);
    userEvent.keyboard(specialChars.arrowRight);
    expect(onInteraction).toHaveBeenCalledWith(false);
    expect(onInteraction).toHaveBeenCalledWith(true);
  });

  test('doesnt allow keyboard when handleKeyboard is false', () => {
    const onInteraction = jest.fn();
    render(
      <ArrowButtons
        value={2}
        maxValue={5}
        minValue={1}
        onInteraction={onInteraction}
      />,
    );

    userEvent.keyboard(specialChars.arrowLeft);
    userEvent.keyboard(specialChars.arrowRight);
    expect(onInteraction).not.toHaveBeenCalled();
  });

  test('doesnt allow using arrow key when value is minValue', () => {
    const onInteraction = jest.fn();
    render(
      <ArrowButtons
        value={1}
        minValue={1}
        maxValue={5}
        onInteraction={onInteraction}
        handleKeyboard={true}
      />,
    );

    userEvent.keyboard(specialChars.arrowLeft);
    expect(onInteraction).not.toHaveBeenCalled();
  });

  test('doesnt allow using arrow key when value is maxValue', () => {
    const onInteraction = jest.fn();
    render(
      <ArrowButtons
        value={5}
        minValue={1}
        maxValue={5}
        onInteraction={onInteraction}
        handleKeyboard={true}
      />,
    );

    userEvent.keyboard(specialChars.arrowRight);
    expect(onInteraction).not.toHaveBeenCalled();
  });
});
