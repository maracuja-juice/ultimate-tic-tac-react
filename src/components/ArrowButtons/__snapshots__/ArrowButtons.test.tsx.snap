// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrowButtons should match snapshot 1`] = `
<Fragment>
  <withRipple(Function)
    className="buttonMargin"
    dense={true}
    disabled={true}
    icon={
      Object {
        "aria-hidden": true,
        "icon": <MdChevronLeft />,
      }
    }
    onClick={[Function]}
    raised={true}
  >
    Previous
  </withRipple(Function)>
  <withRipple(Function)
    className="buttonMargin"
    dense={true}
    disabled={false}
    onClick={[Function]}
    raised={true}
    trailingIcon={
      Object {
        "aria-hidden": true,
        "icon": <MdChevronRight />,
      }
    }
  >
    Next
  </withRipple(Function)>
</Fragment>
`;
